From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement the char_traits::assign forms with a count.

msvcp60's char_traits<char> and char_traits<unsigned short>::assign
with a count, the VC 6 form that takes the character by reference: fills
count characters and returns the pointer.
---
 dlls/msvcp60/msvcp60.spec |  8 ++++----
 dlls/msvcp60/string.c     | 25 +++++++++++++++++++++++++
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index b1a4c23..c0c9632 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -3138,12 +3138,12 @@
 @ cdecl -arch=arm ?assign@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAAAAV12@PBGI@Z(ptr ptr long) MSVCP_basic_string_wchar_assign_cstr_len
 @ thiscall -arch=i386 ?assign@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAEAAV12@PBGI@Z(ptr ptr long) MSVCP_basic_string_wchar_assign_cstr_len
 @ cdecl -arch=win64 ?assign@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QEAAAEAV12@_KG@Z(ptr long long) MSVCP_basic_string_wchar_assignn
-@ stub -arch=win32 ?assign@?$char_traits@D@std@@SAPADPADIABD@Z
-@ stub -arch=win64 ?assign@?$char_traits@D@std@@SAPEADPEAD_KAEBD@Z
+@ cdecl -arch=win32 ?assign@?$char_traits@D@std@@SAPADPADIABD@Z(ptr long ptr) char_traits_char_assignn_ref
+@ cdecl -arch=win64 ?assign@?$char_traits@D@std@@SAPEADPEAD_KAEBD@Z(ptr long ptr) char_traits_char_assignn_ref
 @ cdecl -arch=win32 ?assign@?$char_traits@D@std@@SAXAADABD@Z(ptr ptr) msvcp90.?assign@?$char_traits@D@std@@SAXAADABD@Z
 @ cdecl -arch=win64 ?assign@?$char_traits@D@std@@SAXAEADAEBD@Z(ptr ptr) msvcp90.?assign@?$char_traits@D@std@@SAXAEADAEBD@Z
-@ stub -arch=win32 ?assign@?$char_traits@G@std@@SAPAGPAGIABG@Z
-@ stub -arch=win64 ?assign@?$char_traits@G@std@@SAPEAGPEAG_KAEBG@Z
+@ cdecl -arch=win32 ?assign@?$char_traits@G@std@@SAPAGPAGIABG@Z(ptr long ptr) char_traits_short_assignn_ref
+@ cdecl -arch=win64 ?assign@?$char_traits@G@std@@SAPEAGPEAG_KAEBG@Z(ptr long ptr) char_traits_short_assignn_ref
 @ cdecl -arch=win32 ?assign@?$char_traits@G@std@@SAXAAGABG@Z(ptr ptr) msvcp90.?assign@?$char_traits@G@std@@SAXAAGABG@Z
 @ cdecl -arch=win64 ?assign@?$char_traits@G@std@@SAXAEAGAEBG@Z(ptr ptr) msvcp90.?assign@?$char_traits@G@std@@SAXAEAGAEBG@Z
 @ cdecl -arch=arm ?at@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QAAAADI@Z(ptr long) MSVCP_basic_string_char_at
diff --git a/dlls/msvcp60/string.c b/dlls/msvcp60/string.c
index 253b13b..b4843a4 100644
--- a/dlls/msvcp60/string.c
+++ b/dlls/msvcp60/string.c
@@ -3615,3 +3615,28 @@ void __thiscall basic_string_wchar_insert_iter(basic_string_wchar *this,
     basic_string_wchar_insert_cstr_len(this, basic_string_wchar__Pdif(pos, this->ptr),
             beg, basic_string_wchar__Pdif(end, beg));
 }
+
+/* ?assign@?$char_traits@D@std@@SAPADPADIABD@Z */
+/* ?assign@?$char_traits@D@std@@SAPEADPEAD_KAEBD@Z */
+/* the character passed by reference: num copies of it */
+char* __cdecl char_traits_char_assignn_ref(char *str, size_t num, const char *c)
+{
+    TRACE("(%p %Iu %p)\n", str, num, c);
+    memset(str, *c, num);
+    return str;
+}
+
+/* ?assign@?$char_traits@G@std@@SAPAGPAGIABG@Z */
+/* ?assign@?$char_traits@G@std@@SAPEAGPEAG_KAEBG@Z */
+unsigned short* __cdecl char_traits_short_assignn_ref(unsigned short *str, size_t num, const unsigned short *c)
+{
+    size_t i;
+
+    TRACE("(%p %Iu %p)\n", str, num, c);
+#ifdef SG_MUTANT_ASSIGNN_REF_SHORT
+    if (num) num--;
+#endif
+    for (i = 0; i < num; i++)
+        str[i] = *c;
+    return str;
+}
